src/ContainerGetCapableTrait.php

```diff
@@ -0,0 +1,178 @@
+<?php
+
+namespace Dhii\Data\Container;
+
+use ArrayAccess;
+use Dhii\Util\String\StringableInterface as Stringable;
+use Exception as RootException;
+use InvalidArgumentException;
+use Iterator;
+use OutOfRangeException;
+use Psr\Container\ContainerExceptionInterface;
+use Psr\Container\ContainerInterface;
+use Psr\Container\NotFoundExceptionInterface;
+use Traversable;
+
+/**
+ * Common functionality for reading data from data sets.
+ *
+ * Supported data sets are:
+ * * Arrays
+ * * Objects
+ * * {@link \Psr\Container\ContainerInterface}
+ * * {@link \ArrayAccess}
+ * * {@link \Traversable}
+ *
+ * All SPL data storage classes are supported, given that they almost all implement either {@link \ArrayAccess} or
+ * {@link \Traversable}.
+ *
+ * @since [*next-version*]
+ */
+trait ContainerGetCapableTrait
+{
+    /**
+     * Retrieves a value from a container or data set.
+     *
+     * @since [*next-version*]
+     *
+     * @param array|object|Traversable|ContainerInterface $container The container to read from.
+     * @param string|Stringable                           $key       The key of hte value to retrieve.
+     *
+     * @return mixed The value mapped to the given key.
+     *
+     * @throws ContainerExceptionInterface If an error occurred while reading from the container.
+     * @throws NotFoundExceptionInterface  If the key was not found in the container.
+     */
+    protected function _containerGet($container, $key)
+    {
+        $key = $this->_normalizeString($key);
+        $isContainer = $container instanceof ContainerInterface;
+        $isArrayLike = is_array($container) || $container instanceof ArrayAccess;
+        $isObject = is_object($container);
+        $isTraversable = $container instanceof Traversable;
+
+        if (!$isContainer && !$isArrayLike && !$isObject && !$isTraversable) {
+            throw $this->_createInvalidArgumentException(
+                $this->__('Argument #1 is not a valid container'),
+                null,
+                null,
+                $container
+            );
+        }
+
+        if ($isContainer) {
+            return $container->get($key);
+        }
+
+        if ($isArrayLike && isset($container[$key])) {
+            return $container[$key];
+        }
+
+        if ($isObject && property_exists($container, $key)) {
+            return $container->{$key};
+        }
+
+        if ($isTraversable) {
+            try {
+                $iterator = $this->_resolveIterator($container);
```

Comment:
Not sure if this is actually necessary. You are not requiring the specific `Iterator` API here, nor are you looking for something custom. You are just iterating, and for that `Traversable` is already enough.

Reply:
You're right. I'll make the changes.